src/plugins/es_ui_shared/public/request/use_request.ts

```diff
@@ -0,0 +1,167 @@
+
+  const requestBody = useMemo(() => {
+    return {
+      path,
+      method,
+      query: queryStringified ? query : undefined,
+      body: bodyStringified ? body : undefined,
+    };
+    // queryStringified and bodyStringified stand in for query and body as dependencies.
+    /* eslint-disable-next-line react-hooks/exhaustive-deps */
+  }, [path, method, queryStringified, bodyStringified]);
+
+  const sendRequest = useCallback(async () => {
+    // If we're on an interval, this allows us to reset it if the user has manually requested the
+    // data, to avoid doubled-up requests.
+    clearPollInterval();
+
+    // Pull off request ID and increment request count.
+    const requestId = requestCountRef.current++;
+
+    // We don't clear error or data, so it's up to the consumer to decide whether to display the
+    // "old" error/data or loading state when a new request is in-flight.
+    setIsLoading(true);
+
+    const response = await sendStatelessRequest<D, E>(httpClient, requestBody);
+    const { data: serializedResponseData, error: responseError } = response;
+
+    const isOutdatedRequest = requestId !== requestCountRef.current - 1;
+    const isUnmounted = isMounted.current === false;
+
+    // Ignore outdated or irrelevant data.
+    if (isOutdatedRequest || isUnmounted) {
+      return { data: null, error: null };
+    }
+
+    setError(responseError);
+    // If there's an error, keep the data from the last request in case it's still useful to the user.
+    if (!responseError) {
+      const responseData = deserializer
+        ? deserializer(serializedResponseData)
+        : serializedResponseData;
+      setData(responseData);
+    }
+    // Setting isLoading to false also acts as a semaphore for scheduling the next poll request.
+    setIsLoading(false);
+
+    return { data: serializedResponseData, error: responseError };
+  }, [requestBody, httpClient, deserializer, clearPollInterval]);
+
+  const scheduleRequest = useCallback(() => {
+    // If there's a scheduled poll request, this new one will supersede it.
+    clearPollInterval();
+
+    // Schedule next poll request.
+    if (pollIntervalMs) {
+      pollIntervalIdRef.current = setTimeout(sendRequest, pollIntervalMs);
+    }
+  }, [pollIntervalMs, sendRequest, clearPollInterval]);
+
+  // Send the request on component mount and whenever the dependencies of sendRequest() change.
+  useEffect(() => {
+    sendRequest();
+  }, [sendRequest]);
+
+  // Schedule the next poll request when the previous one completes.
+  useEffect(() => {
+    if (isMounted.current === false) {
+      // Don't schedule a request on component mount.
+      return;
+    }
+
+    // When a request completes, attempt to schedule the next one. Note that we aren't re-scheduling
+    // a request whenever sendRequest's dependencies change.
+    if (!isLoading) {
+      scheduleRequest();
+    }
+  }, [isLoading, scheduleRequest]);
```

Comment:
Note that here we're using `isLoading` to signal the need to reschedule a request, and not `totalRequests` as in https://github.com/cjcenizal/kibana/pull/25/files#diff-a60f7c6c34d3d037970d34856d1b3da2R142. This is so that we don't interfere with the polling rhythm whenever `sendRequest` dependencies change.

Reply:
I think it is a little bit less explicit than with my `totalRequest` count (I personally understand better "when the request count increments, reschedule a request", as "isLoading" _seems_ like it could change independently.) but I am OK to go with your suggestion.

But your comment is not right, the reschedule **will** happen whenever `scheduleRequest` changes, thus whenever its dependencies change.

Also, you are missing a 

```js
if (!isMounted.current) {
  return;
}
```

on top, as we don't want to schedule a request when the component mounts, but when the initial request finished.

Reply:
> But your comment is not right, the reschedule will happen whenever scheduleRequest changes, thus whenever its dependencies change.

Could you explain this a bit? From the way I read this code, it looks like the effect is called whenever `scheduleRequest` (and therefore `sendRequest`) changes, but the `isLoading` condition prevents `scheduleRequest` from being called each time.

> if (!isMounted.current) {
>   return;
> }

`isLoading` is true when the component mounts, so the request isn't being initially scheduled. I'll add a comment to clarify this.

Reply:
For context, I had to switch away from using `totalRequests` because here we're incrementing it before the request completes to get the request ID, instead of after like you originally had. This meant that if we used it to schedule the next request, then the next request would be scheduled before the first one completes.

Reply:
> isLoading is true when the component mounts

This is the problem I think. It should be `false` (as I was expecting it) and it is set to true when `sendRequest()` is called.

> Could you explain this a bit? 

Nevermind I will understand it. For me adding so much comment makes it harder to follow and it seems that there is something "special" the code, but that's ok.
I was saying that the ´useEffect´ **will be** triggered whenever any of the deps changes (which is simply how useEffect works). And of course, the `if` condition prevents the reschedule. But I don't need a comment to explain that.

My mental model was

```js
const [isLoading, setIsLoading] = useState(false); // nothing happened just yet

...

useEffect(() => {
  if (!isMounted.current) {
    return; // clearly indicate to not schedule anything on mount
  }

  if (!isLoading) {
    scheduleRequest();
  }

}, [isLoading, scheduleRequest]);
```
With the above code, I don't need any comment as it is self explanatory.

> For context, I had to switch away from using totalRequests because here we're incrementing it before the request completes to get the request ID, instead of after like you originally suggested. This meant that if we used it to schedule the next request, then the next request would be scheduled before the first one completes.

I am not following but that's OK I guess 😄